src/runtime-rs/crates/runtimes/src/manager.rs

```diff
@@ -434,14 +450,67 @@ impl RuntimeHandlerManager {
 
             Ok(TaskResponse::CreateContainer(shim_pid))
         } else {
-            self.handler_request(req)
+            self.handler_task_request(req)
                 .await
                 .context("handler TaskRequest")
         }
     }
 
+    pub async fn handler_sandbox_request(&self, req: SandboxRequest) -> Result<SandboxResponse> {
+        let instance = self
+            .get_runtime_instance()
+            .await
+            .context("get runtime instance")?;
+        let sandbox = instance.sandbox.clone();
+
+        match req {
+            SandboxRequest::CreateSandbox(req) => Err(anyhow!("Unreachable request {:?}", req)),
+            SandboxRequest::StartSandbox(_) => {
+                sandbox
+                    .start()
+                    .await
+                    .context("start sandbox in sandbox handler")?;
+                Ok(SandboxResponse::StartSandbox(StartSandboxInfo {
+                    pid: std::process::id(),
+                    create_time: Some(SystemTime::now()),
+                }))
+            }
+            SandboxRequest::Platform(_) => Ok(SandboxResponse::Platform(PlatformInfo {
+                os: std::env::consts::OS.to_string(),
+                architecture: std::env::consts::ARCH.to_string(),
+            })),
+            SandboxRequest::StopSandbox(_) => {
+                sandbox.stop().await.context("stop sandbox")?;
+
+                Ok(SandboxResponse::StopSandbox)
+            }
+            SandboxRequest::WaitSandbox(_) => {
+                let exit_info = sandbox.wait().await.context("wait sandbox")?;
+
+                Ok(SandboxResponse::WaitSandbox(exit_info))
+            }
+            SandboxRequest::SandboxStatus(_) => {
+                let status = sandbox.status().await?;
```

Comment:
What's the purpose of the `info` field in this struct (`common::sandbox::SandboxStatus`)?  It's immediately dropped here when `status` is translated into `SandboxStatusInfo`, and `VirtSandbox`'s `Sandbox::status()` implementation (elsewhere in this commit) doesn't even try to fill it in to begin with.

Also, more generally, this commit seems define groups of similar structures and translations among them, like for instance here - sandbox status is first produced in a `Sandbox::status()` implementation as `common::sandbox::SandboxStatus`.  It's then translated to `SandboxStatusInfo` here in `handler_sandbox_request()` which in turn seems translated into `SandboxStatusResponse` (by a `TryFrom` implementation in `trans_into_shim.rs`).  All of the structures involved appear largely the same.

I do understand that we might want separation between sandbox api protocol and internal implementation of the shim which would justify the latter translation.  But what's the purpose of the first one?

Reply:
Hi @pmores 

Yes, you are right. The reason why several status structures seem to be repeatedly defined here is to make the implementation on the kata runtime side independent of the protocol on the containerd side. Therefore, it is necessary to convert between the definition of the kata runtime and the definition on the containerd side.

In addition, the info field here is to align with the field defined by containerd: https://github.com/containerd/rust-extensions/blob/main/crates/shim-protos/vendor/github.com/containerd/containerd/api/runtime/sandbox/v1/sandbox.proto#L120. However, I see that this field is not actually used yet, so I did not assign the content of this field, but dropped it directly.

Reply:
I do agree that separating shim internal implementation from network protocol handling is an excellent idea.  

Still, I would hope that we could do without `SandboxStatusInfo` (the middle representation referring to my previous comment), at least in principle, however I can also see that `SandboxStatusInfo`  appears to be part of a larger idea that I might not be understanding or appreciating properly.

Specifically, it seems that in between the network protocol implementation (`SandboxStatusResponse` in the sandbox status use case) and shim internal implementation (`common::sandbox::SandboxStatus`) there's an intervening layer which `SandboxStatusInfo` seems to be part of and which looks like the sandbox api protocol representation, _however_ freed from dependency on containerd.

Is that the idea, or am I getting it wrong?  Could you comment on the purpose of this layer?  (I'm not disagreeing or criticising at this point, for now I just want to make sure I understand correctly.)

Reply:
That's right, our original design purpose was to consider the following layers: the first is the containerd GRPC network protocol, the middle layer is the implementation of kata shim, and the other layer is the specific sandbox implementation layer. We want the implementation of the middle layer kata shim, that is, the implementation of the sandbox api, to be independent of the definition of containerd GRPC. So this layer introduces the implementation of the SandboxStatusInfo layer, for example. The sandbox api layer is a general layer. We want it to support VM-based sandboxes as well as WASM-based or runC-based sandboxes. The SandboxStatus layer is a specific sandbox implementation. The reason is that we want the specific sandbox implementation to be independent of the sandbox api implementation of kata shim.

Reply:
Okay, makes sense, thanks!